src/systems/variational_smoother_system.C

```diff
@@ -113,90 +114,135 @@ void VariationalSmootherSystem::prepare_for_smoothing()
   std::map<ElemType, std::vector<Real>> target_elem_inverse_jacobian_dets;
 
   for (const auto * elem : mesh.active_local_element_ptr_range())
-  {
-    femcontext.pre_fe_reinit(*this, elem);
-    femcontext.elem_fe_reinit();
-
-    // Add target element info, if applicable
-    if (_target_inverse_jacobians.find(elem->type()) == _target_inverse_jacobians.end())
     {
-      // Create FEMap to compute target_element mapping information
-      FEMap fe_map_target;
-
-      // pre-request mapping derivatives
-      const auto & dxyzdxi = fe_map_target.get_dxyzdxi();
-      const auto & dxyzdeta = fe_map_target.get_dxyzdeta();
-      // const auto & dxyzdzeta = fe_map_target.get_dxyzdzeta();
-
-      const auto & qrule_points = femcontext.get_element_qrule().get_points();
-      const auto & qrule_weights = femcontext.get_element_qrule().get_weights();
-      const auto nq_points = femcontext.get_element_qrule().n_points();
+      femcontext.pre_fe_reinit(*this, elem);
+      femcontext.elem_fe_reinit();
 
-      // If the target element is the reference element, Jacobian matrix is
-      // identity, det of inverse is 1. This will only be overwritten if a
-      // different target elemen is explicitly specified.
-      target_elem_inverse_jacobian_dets[elem->type()] =
-          std::vector<Real>(nq_points, 1.0);
-
-      switch (elem->type())
+      // Add target element info, if applicable
+      if (_target_inverse_jacobians.find(elem->type()) == _target_inverse_jacobians.end())
         {
-          case TRI3:
+          // Create FEMap to compute target_element mapping information
+          FEMap fe_map_target;
+
+          // pre-request mapping derivatives
+          const auto & dxyzdxi = fe_map_target.get_dxyzdxi();
+          const auto & dxyzdeta = fe_map_target.get_dxyzdeta();
+          // const auto & dxyzdzeta = fe_map_target.get_dxyzdzeta();
+
+          const auto & qrule_points = femcontext.get_element_qrule().get_points();
+          const auto & qrule_weights = femcontext.get_element_qrule().get_weights();
+          const auto nq_points = femcontext.get_element_qrule().n_points();
+
+          // If the target element is the reference element, Jacobian matrix is
+          // identity, det of inverse is 1. This will only be overwritten if a
+          // different target elemen is explicitly specified.
+          target_elem_inverse_jacobian_dets[elem->type()] =
+              std::vector<Real>(nq_points, 1.0);
+
+          // Elems deriving from Tri
+          if (dynamic_cast<const Tri*>(elem))
```

Comment:
Let's rely on the switch statement here?  I know it's not an inner kernel and so this might be the most appropriate idiom but I still want to avoid non-optional RTTI.